virttest/libvirt_xml/vm_xml.py

```diff
@@ -677,14 +677,22 @@ def undefine(self, options=None, virsh_instance=base.virsh):
 
         return virsh_instance.remove_domain(self.vm_name, options)
 
-    def define(self, virsh_instance=base.virsh):
-        """Define VM with virsh from this instance"""
+    def define(self, virsh_instance=base.virsh, detail=False):
+        """
+        Define VM with virsh from this instance
+
+        :param virsh_instance: the virsh instance to be operated with
+        :param detail: whether return command result object when define failed
+        :return: True if succeeded,
+                 False if failed and detail is not required,
+                 cmd result object if failed and detail is required.
+        """
         result = virsh_instance.define(self.xml)
         if result.exit_status:
-            logging.debug("Define %s failed.\n"
+            logging.error("Define %s failed.\n"
                           "Detail: %s.", self.vm_name,
                           result.stderr_text)
-            return False
+            return result if detail else False
```

Comment:
It is not advisable to return different data type as return, anyway we print error here itself, so why to return the result, instead we can preserve old behaviour for return..

Reply:
This is the error message after this commit:
```
ERROR 1-type_specific.local.virsh.event.positive_test.qemu_monitor_event.pretty_option.device-removal-failed_event.no_timeout.loop -> LibvirtXMLError: Failed to define avocado-vt-vm1 for reason:
error: Failed to define domain from /tmp/xml_utils_temp_e8cturva.xml
error: XML error: Exactly one 'cell' element per guest NUMA cell allowed, non-contiguous ranges or ranges not starting from 0 are not allowed

```

This is before :
```
ERROR 1-type_specific.local.virsh.event.positive_test.qemu_monitor_event.pretty_option.device-removal-failed_event.no_timeout.loop -> LibvirtXMLError: Failed to define avocado-vt-vm1 from xml:
<?xml version='1.0' encoding='UTF-8'?>
<domain type="kvm">
  <name>avocado-vt-vm1</name>
  <uuid>1c9c5d62-4dfa-4778-85ed-6dce62b7b518</uuid>
  <metadata>
    <ns0:libosinfo xmlns:ns0="http://libosinfo.org/xmlns/libvirt/domain/1.0">
      <ns0:os id="http://redhat.com/rhel/8.3" />
    </ns0:libosinfo>
  </metadata>
  <memory unit="KiB">4048896</memory>
  <currentMemory unit="KiB">4048896</currentMemory>
  <vcpu placement="static">2</vcpu>
  <resource>
    <partition>/machine</partition>
  </resource>
  <os>
    <type arch="x86_64" machine="pc-q35-rhel8.3.0">hvm</type>
    <boot dev="hd" />
  </os>
  <features>
    <acpi />
    <apic />
  </features>
  <clock offset="utc">
    <timer name="rtc" tickpolicy="catchup" />
    <timer name="pit" tickpolicy="delay" />
    <timer name="hpet" present="no" />
  </clock>
  <on_poweroff>destroy</on_poweroff>
  <on_reboot>restart</on_reboot>
  <on_crash>destroy</on_crash>
  <pm>
    <suspend-to-mem enabled="no" />
    <suspend-to-disk enabled="no" />
  </pm>
  <devices>
    <emulator>/usr/libexec/qemu-kvm</emulator>
    <disk device="disk" type="file">
      <driver name="qemu" type="qcow2" />
      <source file="/var/lib/avocado/data/avocado-vt/images/jeos-27-x86_64.qcow2" />
      <backingStore />
      <target bus="virtio" dev="vda" />
      <address bus="0x04" domain="0x0000" function="0x0" slot="0x00" type="pci" />
    </disk>
    <controller index="0" model="qemu-xhci" ports="15" type="usb">
      <address bus="0x02" domain="0x0000" function="0x0" slot="0x00" type="pci" />
    </controller>
    <controller index="0" type="sata">
      <address bus="0x00" domain="0x0000" function="0x2" slot="0x1f" type="pci" />
    </controller>
    <controller index="0" model="pcie-root" type="pci" />
    <controller index="1" model="pcie-root-port" type="pci">
      <model name="pcie-root-port" />
      <target chassis="1" port="0x10" />
      <address bus="0x00" domain="0x0000" function="0x0" multifunction="on" slot="0x02" type="pci" />
    </controller>
    <controller index="2" model="pcie-root-port" type="pci">
      <model name="pcie-root-port" />
      <target chassis="2" port="0x11" />
      <address bus="0x00" domain="0x0000" function="0x1" slot="0x02" type="pci" />
    </controller>
    <controller index="3" model="pcie-root-port" type="pci">
      <model name="pcie-root-port" />
      <target chassis="3" port="0x12" />
      <address bus="0x00" domain="0x0000" function="0x2" slot="0x02" type="pci" />
    </controller>
    <controller index="4" model="pcie-root-port" type="pci">
      <model name="pcie-root-port" />
      <target chassis="4" port="0x13" />
      <address bus="0x00" domain="0x0000" function="0x3" slot="0x02" type="pci" />
    </controller>
    <controller index="5" model="pcie-root-port" type="pci">
      <model name="pcie-root-port" />
      <target chassis="5" port="0x14" />
      <address bus="0x00" domain="0x0000" function="0x4" slot="0x02" type="pci" />
    </controller>
    <controller index="6" model="pcie-root-port" type="pci">
      <model name="pcie-root-port" />
      <target chassis="6" port="0x15" />
      <address bus="0x00" domain="0x0000" function="0x5" slot="0x02" type="pci" />
    </controller>
    <controller index="7" model="pcie-root-port" type="pci">
      <model name="pcie-root-port" />
      <target chassis="7" port="0x16" />
      <address bus="0x00" domain="0x0000" function="0x6" slot="0x02" type="pci" />
    </controller>
    <controller index="0" type="virtio-serial">
      <address bus="0x03" domain="0x0000" function="0x0" slot="0x00" type="pci" />
    </controller>
    <interface type="network">
      <mac address="52:54:00:6a:3b:13" />
      <source network="default" />
      <model type="virtio" />
      <address bus="0x01" domain="0x0000" function="0x0" slot="0x00" type="pci" />
    </interface>
    <serial type="pty">
      <target port="0" type="isa-serial">
        <model name="isa-serial" />
      </target>
    </serial>
    <console type="pty">
      <target port="0" type="serial" />
    </console>
    <channel type="unix">
      <target name="org.qemu.guest_agent.0" type="virtio" />
      <address bus="0" controller="0" port="1" type="virtio-serial" />
    </channel>
    <input bus="usb" type="tablet">
      <address bus="0" port="1" type="usb" />
    </input>
    <input bus="ps2" type="mouse" />
    <input bus="ps2" type="keyboard" />
    <graphics autoport="yes" listen="127.0.0.1" port="-1" type="vnc">
      <listen address="127.0.0.1" type="address" />
    </graphics>
    <video>
      <model heads="1" primary="yes" ram="65536" type="qxl" vgamem="16384" vram="65536" />
      <address bus="0x00" domain="0x0000" function="0x0" slot="0x01" type="pci" />
    </video>
    <memballoon model="virtio">
      <address bus="0x05" domain="0x0000" function="0x0" slot="0x00" type="pci" />
    </memballoon>
    <rng model="virtio">
      <backend model="random">/dev/urandom</backend>
      <address bus="0x06" domain="0x0000" function="0x0" slot="0x00" type="pci" />
    </rng>
  </devices>
  <seclabel model="selinux" relabel="yes" type="dynamic" />
  <seclabel model="dac" relabel="yes" type="dynamic" />
<maxMemory slots="16" unit="KiB">25600000</maxMemory><cpu><numa><cell cpus="0-1" id="2" memory="4048896" unit="KiB" /></numa></cpu></domain>
```

When we collect and analyze error message, we would like more detailed reason, not just xml content. And as you can see from the original error message, there is no reason of why define failed. 

For  `so why to return the result, instead we can preserve old behaviour for return..` , if define() doesn't return result object, I'll have to run 'virsh define' again to get the output, and put it in error message of function sync(). Therefore I choose to modify the define() function instead of running another 'virsh define'

Reply:
@sathnaga Could you help review again? Thanks

Reply:
@chloerh I understand your concern, I meant by `result.stderr_text` is printed here in this method,which is what you  are expecting as a failure message for user, and I agree, it is convenient for user debugging to print that error message in test fail msg, which is your new change in other place is doing, all my concern is having different type of return for the method, if we can choose to send both like tuple( True, result) and (False, result) or result alone and let other place decide it has failed or passed, am fine with.

Hope this helps..

Reply:
Running 'virsh define' again:
```
        if not self.define(virsh_instance=virsh_instance):
            result_define = virsh_instance.define(self.xml)
            if backup:
                backup.define(virsh_instance=virsh_instance)
            logging.error("Failed to define %s from xml:\n%s"
                          % (self.vm_name, self.xmltreefile))
            raise xcepts.LibvirtXMLError("Failed to define %s for reason:\n%s"
                                         % (self.vm_name, result_define.stderr_text))
```
Could avoid modifying function `define()`

How about this way? @sathnaga

Reply:
@chloerh this would be little repeat of the same action again, am fine with if you would like to not disturb too much `define()` method, following tp-libvirt instances uses  that define method.

if these two below statements are going to get us same behaviour
```
self.define(virsh_instance=virsh_instance):
result_define = virsh_instance.define(self.xml)
```
can we use only `result_define = virsh_instance.define(self.xml)` and check for failure from `result_define` and proceed ?


```
libvirt/tests/src/serial/serial_functional.py:        vm_xml.undefine()
libvirt/tests/src/serial/serial_functional.py:        res = vm_xml.virsh.define(vm_xml.xml)
libvirt/tests/src/serial/serial_functional.py:            vm_xml.undefine()
libvirt/tests/src/controller/multiple_controller_ppc.py:        vm_xml.undefine()
libvirt/tests/src/controller/multiple_controller_ppc.py:        res = vm_xml.virsh.define(vm_xml.xml)
libvirt/tests/src/controller/controller_functional.py:        res = vm_xml.virsh.define(guest_xml.xml)
libvirt/tests/src/virsh_cmd/domain/virsh_migrate_vm_cfg.py:            vm_xml_backup.define()
libvirt/tests/src/virsh_cmd/domain/virsh_migrate_vm_cfg.py:            vm_xml_backup.define()
libvirt/tests/src/channel/channel_functional.py:        vm_xml.undefine()
libvirt/tests/src/channel/channel_functional.py:        res = vm_xml.virsh.define(vm_xml.xml)
libvirt/tests/src/virtual_device/tpm_device.py:                                vm_xml.define()
```

Reply:
> @chloerh this would be little repeat of the same action again, am fine with if **you would like to not disturb too much `define()` method, following tp-libvirt instances uses that define method.**
Yes, that's what I was thinking... Thank you for understanding me.
> 
> ```
> libvirt/tests/src/serial/serial_functional.py:        vm_xml.undefine()
> libvirt/tests/src/serial/serial_functional.py:        res = vm_xml.virsh.define(vm_xml.xml)
> libvirt/tests/src/serial/serial_functional.py:            vm_xml.undefine()
> libvirt/tests/src/controller/multiple_controller_ppc.py:        vm_xml.undefine()
> libvirt/tests/src/controller/multiple_controller_ppc.py:        res = vm_xml.virsh.define(vm_xml.xml)
> libvirt/tests/src/controller/controller_functional.py:        res = vm_xml.virsh.define(guest_xml.xml)
> libvirt/tests/src/virsh_cmd/domain/virsh_migrate_vm_cfg.py:            vm_xml_backup.define()
> libvirt/tests/src/virsh_cmd/domain/virsh_migrate_vm_cfg.py:            vm_xml_backup.define()
> libvirt/tests/src/channel/channel_functional.py:        vm_xml.undefine()
> libvirt/tests/src/channel/channel_functional.py:        res = vm_xml.virsh.define(vm_xml.xml)
> libvirt/tests/src/virtual_device/tpm_device.py:                                vm_xml.define()
> ```
> 
> if these two below statements are going to get us same behaviour
> 
> ```
> self.define(virsh_instance=virsh_instance):
> result_define = virsh_instance.define(self.xml)
> ```
> 
> can we use only `result_define = virsh_instance.define(self.xml)` and check for failure from `result_define` and proceed ?

You are right! I didn't think of that. I was stuck in the circle of 'not changing too much code' .... anyway, Thank you for the good suggestion, I'm going to take it.

Reply:
@sathnaga I have updated code and run a few tests to verify. Could you help review again? Thanks.